spec/validations_spec.cr

```diff
@@ -18,6 +18,60 @@ private def attribute(value)
 end
 
 describe Avram::Validations do
+  describe "validate_at_most_one_filled" do
+    it "marks filled attribute as invalid if more than one is filled" do
+      filled_attribute = attribute("filled")
+      filled_attribute_2 = attribute("filled")
+      blank_attribute = attribute("")
+      Avram::Validations.validate_at_most_one_filled(filled_attribute, filled_attribute_2, blank_attribute)
+      filled_attribute.valid?.should be_true
+      blank_attribute.valid?.should be_true
+      filled_attribute_2.errors.should eq(["must be blank"])
+    end
+
+    it "does not mark any fields as invalid if just one is filled" do
+      filled_attribute = attribute("filled")
+      blank_attribute = attribute("")
+
+      Avram::Validations.validate_at_most_one_filled(filled_attribute, blank_attribute)
+
+      filled_attribute.valid?.should be_true
+      blank_attribute.valid?.should be_true
+    end
+  end
+
+  describe "validate_exactly_one_filled" do
+    it "marks filled attribute as invalid if more than one is filled" do
+      filled_attribute = attribute("filled")
+      filled_attribute_2 = attribute("filled")
+      blank_attribute = attribute("")
+      Avram::Validations.validate_exactly_one_filled(filled_attribute, filled_attribute_2, blank_attribute)
+      filled_attribute.valid?.should be_true
+      blank_attribute.valid?.should be_true
+      filled_attribute_2.errors.should eq(["must be blank"])
+    end
+
+    it "marks first field as invalid if no attributes are filled" do
+      first_blank_attribute = attribute(nil)
+      second_blank_attribute = attribute("")
```

Comment:
Just kind of riffing off this in my head, it might be cool to be able to set what it counts as `blank` under the hood, so you could do things like:

```crystal
tos_accepted = attribute("0")
just_browsing = attribute("1")

Avram::Validations.validate_at_most_one_filled(...
```

Reply:
This actually is what happens. Since attribute.value is used it will call blank on the parsed type. So if you have a Boolean it’ll use blank on that. Does that make sense?

Reply:
It does, but I think I explained my idea poorly. I'm saying that it might parse the value, get `false`, assume that `false.blank?` is "not filled" but you may not want that. However, trying to think of an example I can't actually think of anything that can't be solved better with a custom validation. Here's my contrived example anyway:

```crystal
product_image = attribute("/path/to/default.png") # no imaged uploaded, blank? => false
no_image_checkbox = attribute("0") # evaluates to `false`, blank? => true

validate_exactly_one_filled(product_image, no_image_checkbox) # passes, but shouldn't because there's no image uploaded and the `no image checkbox` wasn't checked.
```

So I'm saying you could define whatever predicate that lets a field pass or fail this validation for the `product_image` field to be `!= "/path/to/default.png"`.

Reply:
I think I’m that case you’d want to create some kind of custom type. Like a custom Boolean or a FilePath type so you can change what blank means. That way it changes what blank? Is everywhere. Like most things, creating custom types for columns needs to be documented 😂

Reply:
Gotcha, that makes sense and is a better solution!